pallets/claim/src/lib.rs

```diff
@@ -0,0 +1,467 @@
+            /// How many tokens were claimed
+            amount: BalanceOf<T>,
+            /// The payment id
+            payment_id: <T::Paymaster as Pay>::Id,
+        },
+        /// Ending of the airdrop campaign
+        AirdropEnded {
+            /// The id of the airdrop that has just ended
+            airdrop_id: u64,
+        },
+    }
+
+    /// Error for the claim pallet.
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Attemp to start a new airdrop while there is one already in progress
+        AlreadyStarted,
+        /// The pot has not enough funds available to cover for all the airdropped amounts
+        NotEnoughFunds,
+        /// Account requested a claim but it is not present among the Beneficiaries
+        NotEligible,
+        /// Added a beneficiary without balance to claim
+        NothingToClaim,
+        /// Attempt to modify the balance of an already added beneficiary
+        AlreadyPresent,
+        /// There was some issue with the mechanism of payment.
+        PayoutError,
+        /// Attempt to perform an action implying an open airdrop, while it has already ended
+        AlreadyEnded,
+    }
+
+    impl<T: Config> Pallet<T> {
+        /// The account ID of the claim pot.
+        ///
+        /// This actually does computation. If you need to keep using it, then make sure you cache the
+        /// value and only call this once.
+        pub fn account_id() -> T::AccountId {
+            T::PalletId::get().into_account_truncating()
+        }
+
+        /// Return the amount of money in the pot.
+        /// The existential deposit is not part of the pot so claim account never gets deleted.
+        pub fn pot() -> BalanceOf<T> {
+            T::Currency::reducible_balance(
+                &Self::account_id(),
+                Preservation::Preserve,
+                Fortitude::Polite,
+            )
+        }
+
+        fn do_claim(origin: T::AccountId, beneficiary: Option<T::AccountId>) -> DispatchResult {
+            // See if account is eligible to get an airdrop
+            Beneficiaries::<T>::try_mutate_exists(origin.clone(), |amount| {
+                *amount = match amount {
+                    // Account is eligible to get an airdrop
+                    Some(amount) => {
+                        // Determine who is the beneficiary
+                        let beneficiary = beneficiary.unwrap_or(origin);
+                        // Execute payment
+                        let available = Self::pot();
+                        if *amount > available {
+                            log::warn!("Claimable amount {amount:?} bigger than total available {available:?}");
+                            Err(Error::<T>::PayoutError)?; // Prevent going under the existential deposit of the account
+                        }
+                        let payment_id = T::Paymaster::pay(&beneficiary, (), *amount)
+                            .map_err(|_| Error::<T>::PayoutError)?;
```

Comment:
Again, maybe is better to remove `Paymaster` abstraction because here you didn't have any assurance that the Paymaster have enough founds.